haystack/preview/components/reader.py

```diff
@@ -0,0 +1,151 @@
+    ) -> Tuple[List[List[int]], List[List[int]], List[List[float]]]:
+        start = output.start_logits
+        end = output.end_logits
+        mask = sequence_ids == 1
+        mask[..., 0] = True
+        mask = torch.logical_and(mask, attention_mask == 1)
+        start = torch.where(mask, start, -torch.inf)
+        end = torch.where(mask, end, -torch.inf)
+        start = torch.softmax(start, -1)
+        end = torch.softmax(end, -1)
+        start = start.unsqueeze(-1)
+        end = end.unsqueeze(-2)
+
+        probabilities = start * end  # shape: (batch_size, seq_length (start), seq_length (end))
+        masked_probabilities = torch.triu(probabilities, diagonal=1)
+        masked_probabilities[..., 0, 0] = probabilities[..., 0, 0]
+        masked_probabilities[..., 0, 1:] = 0
+        flat_probabilities = masked_probabilities.flatten(-2, -1)  # necessary for topk
+        candidates = torch.topk(flat_probabilities, top_k)
+        start_candidates = candidates.indices // max_seq_length
+        end_candidates = candidates.indices % max_seq_length
+        start_candidates = start_candidates.cpu()
+        end_candidates = end_candidates.cpu()
+
+        start_candidates = [
+            [encoding.token_to_chars(start)[0] if start != 0 else 0 for start in candidates]
+            for candidates, encoding in zip(start_candidates, encodings)
+        ]
+        end_candidates = [
+            [encoding.token_to_chars(end)[1] if end != 0 else 0 for end in candidates]
+            for candidates, encoding in zip(end_candidates, encodings)
+        ]
+        probabilities = candidates.values.cpu().tolist()
+
+        return start_candidates, end_candidates, probabilities
+
+    def _unflatten(
+        self,
+        start: List[List[int]],
+        end: List[List[int]],
+        probabilities: List[List[float]],
+        flattened_documents: List[str],
+        documents: List[List[Document]],
+        top_k: int,
+    ) -> List[List[Answer]]:
+        answers = [
+            [document[start:end] for start, end in zip(start_candidates_, end_candidates_)]
+            for document, start_candidates_, end_candidates_ in zip(flattened_documents, start, end)
+        ]
+
+        i = 0
+        nested_answers = []
+        for documents_ in documents:
+            current_answers = []
+            j = i + len(documents_)
+            for answers_, probabilities_ in zip(answers[i:j], probabilities[i:j]):
+                for answer, probability in zip(answers_, probabilities_):
+                    current_answers.append(Answer(answer=answer, probability=probability))
+            i = j
+            current_answers = sorted(current_answers, key=lambda answer: answer.probability, reverse=True)[:top_k]
+            nested_answers.append(current_answers)
+
+        return nested_answers
+
+    @component.output_types(answers=List[List[Answer]])
+    def run(
+        self,
+        queries: List[str],
+        documents: List[List[Document]],
```

Comment:
Is this already a batch or should a batch size be specified separately so the method can perform additional batching?

Reply:
My idea is that this is already batched. Are there constraints on the batch size that would make this approach not viable? Do you need to control the batch size in any way?

Reply:
Maybe then we can have a "max batch size": if the incoming batch is larger, we split it to fit. It's a bit of a mess to juggle all these batches but it should be possible. Does that sound alright?

Reply:
That sounds good and it wasn't too complex to add.